check.go

```diff
@@ -0,0 +1,286 @@
+		c.Failed = true
+		return
+	}
+
+	if c.Shell != "" && c.Web != "" {
+		fmt.Fprintf(os.Stderr, "[%s] Web and shell checks in one block are not allowed.\n", c.Name)
+		c.Failed = true
+		return
+	}
+
+	// Default timeout.
+	if c.Repeat == 0 {
+		c.Repeat = 30
+	}
+
+	// Default to 1 attempt.
+	if c.Tries == 0 {
+		c.Tries = 1
+	}
+
+	// Default successful http return code
+	if c.Return == 0 {
+		c.Return = 200
+	}
+
+	r := time.Duration(c.Repeat) * time.Second
+	s := time.Duration(c.Sleep) * time.Second
+
+	regex, err := regexp.Compile(c.Match)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "[%s] Invalid regular expression: %s\n", c.Name, err.Error())
+		return
+	}
+
+	var check checkFunc
+	if c.Web != "" {
+		if c.Name == "" {
+			c.Name = c.Web
+		}
+		check = c.runWeb
+	} else {
+		if c.Name == "" {
+			c.Name = c.Shell
+		}
+		check = c.runShell
+	}
+
+	for {
+		subject, msg := check(s, regex)
+
+		if subject != "" {
+			c.Since = time.Now().Format(time.RFC3339)
+
+			fmt.Fprintf(os.Stderr, "[%s] %s", c.Name, msg)
+
+			if c.Notify != "" {
+				go c.notify(subject, msg)
+			}
+			if c.Alert != "" {
+				go c.alert(msg, c.Failed)
+			}
+		}
+
+		time.Sleep(r)
+	}
+}
+
+func (c *Check) runWeb(s time.Duration, r *regexp.Regexp) (subject string, msg string) {
+	fmt.Printf("[%s] Running web check\n", c.Name)
+
+	err := c.fetch(c.Web, c.Return, r)
+	for i := 1; err != nil && i < c.Tries; i++ {
+		time.Sleep(s)
+		fmt.Printf("[%s] Running web check, retry #%d\n", c.Name, i)
+		err = c.fetch(c.Web, c.Return, r)
+	}
+
+	c.m.Lock()
+	if err != nil {
+		if !c.Failed {
+			c.Failed = true
+			msg = err.Error()
+			subject = "Failed: " + c.Name
+		}
+	} else {
+		if c.Failed {
+			c.Failed = false
+			subject = "Fixed: " + c.Name
+		}
+	}
+	c.m.Unlock()
+
+	return subject, msg
+}
+
+func (c *Check) runShell(s time.Duration, r *regexp.Regexp) (subject string, msg string) {
+	fmt.Printf("[%s] Running shell check\n", c.Name)
+
+	// Execute with shell with limited attempts
+	out, err := c.execute(c.Shell, r)
+	for i := 1; err != nil && i < c.Tries; i++ {
+		time.Sleep(s)
+		fmt.Printf("[%s] Running shell check, retry #%d\n", c.Name, i)
+		out, err = c.execute(c.Shell, r)
+	}
+
+	if err == nil && r != nil && !r.Match(out) {
+		err = errors.New("Expected:\n" + c.Match + "\n\nGot:\n" + string(out))
+	}
+
+	c.m.Lock()
+	if err != nil {
+		if !c.Failed {
+			c.Failed = true
+			msg = string(out) + err.Error()
+			subject = "Failed: " + c.Name
+		}
+	} else {
+		if c.Failed {
+			c.Failed = false
+			subject = "Fixed: " + c.Name
+		}
+	}
+	c.m.Unlock()
+
+	return subject, msg
+}
+
+func (c *Check) fetch(url string, code int, r *regexp.Regexp) error {
+	resp, err := http.Get(url)
+
+	defer func() {
+		if resp != nil {
+			resp.Body.Close()
+		}
+	}()
+
+	if err != nil {
+		return err
+	}
+
+	// Check status code.
+	if resp.StatusCode != code {
+		return errors.New(url + " returned " + strconv.Itoa(resp.StatusCode))
```

Comment:
Since we are using jsonmon within docker containers we have the requirement to stdout and stderr. So maybe we add an option to handle the logging driver. Maybe this could be default. Nice to have would be another option to wrap this as json log line but this could be another feature PR. Since this application is a small one why so many log types? Are you using this in combination with syslog?